File: src/main/java/com/chattriggers/ctjs/utils/config/ConfigBoolean.java
package com.chattriggers.ctjs.utils.config;

import com.chattriggers.ctjs.minecraft.libs.ChatLib;
import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer;
import com.chattriggers.ctjs.minecraft.wrappers.Client;
import net.minecraft.client.gui.GuiButton;

import javax.annotation.Nullable;

public class ConfigBoolean extends ConfigOption {
    public Boolean value = null;
    private Boolean defaultValue;

    private transient GuiButton button;

    ConfigBoolean(@Nullable ConfigBoolean previous, String name, Boolean defaultValue, int x, int y) {
        super(ConfigOption.Type.BOOLEAN);

        this.name = name;
        this.defaultValue = defaultValue;

        if (previous == null)
            this.value = this.defaultValue;
        else
            this.value = previous.value;

        this.x = x;
        this.y = y;
    }

    private String getStringValue() {
        if (this.value)
            return ChatLib.addColor("&aTrue");
        return ChatLib.addColor("&cFalse");
    }

    @Override
    public void init() {
        super.init();

        this.button = new GuiButton(
                0,
                Renderer.getRenderWidth() / 2 - 100 + this.x,
                this.y + 15,
                getStringValue()
        );
    }

    @Override
    public void draw(int mouseX, int mouseY) {
        if (this.hidden) return;

        int middle = Renderer.getRenderWidth() / 2;

        Renderer.rectangle(0x80000000, middle - 105 + this.x, this.y - 5, 210, 45)
                .setShadow(0xd0000000, 3, 3)
                .draw();

        Renderer.text(this.name, middle - 100 + this.x, this.y).draw();

        this.button.xPosition = middle - 100 + this.x;
        this.button.drawButton(Client.getMinecraft(), mouseX, mouseY);

        super.draw(mouseX, mouseY);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (this.hidden) return;

        if (this.button.mousePressed(Client.getMinecraft(), mouseX, mouseY)) {
            this.value = !this.value;
            this.button.playPressSound(Client.getMinecraft().getSoundHandler());
        }

        if (this.resetButton.mousePressed(Client.getMinecraft(), mouseX, mouseY)) {
            this.value = this.defaultValue;
            this.resetButton.playPressSound(Client.getMinecraft().getSoundHandler());
        }

        this.button.displayString = getStringValue();
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/utils/config/ConfigBoolean.kt
package com.chattriggers.ctjs.utils.config

import com.chattriggers.ctjs.minecraft.libs.ChatLib
import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer
import com.chattriggers.ctjs.minecraft.wrappers.Client
import net.minecraft.client.gui.GuiButton

class ConfigBoolean internal constructor(previous: ConfigBoolean? = null, name: String = "", private val defaultValue: Boolean = false, x: Int = 0, y: Int = 0) : ConfigOption(ConfigOption.Type.BOOLEAN) {
    var value: Boolean = previous?.value ?: this.defaultValue

    @Transient
    private var button: GuiButton = GuiButton(0,
            Renderer.screen.getWidth() / 2 - 100 + this.x, this.y + 15,
            stringValue)

    private val stringValue: String
        get() = if (this.value) ChatLib.addColor("&aTrue") else ChatLib.addColor("&cFalse")

    init {
        this.name = name

        this.x = x
        this.y = y
    }

    override fun draw(mouseX: Int, mouseY: Int, partialTicks: Float) {
        if (this.hidden) return

        val middle = Renderer.screen.getWidth() / 2

        Renderer.rectangle(-0x80000000, (middle - 105 + this.x).toFloat(), (this.y - 5).toFloat(), 210f, 45f)
                .setShadow(-0x30000000, 3f, 3f)
                .draw()

        Renderer.text(this.name!!, (middle - 100 + this.x).toFloat(), this.y.toFloat()).draw()

        //#if MC<=10809
        this.button.xPosition = middle - 100 + this.x
        this.button.drawButton(Client.getMinecraft(), mouseX, mouseY)
        //#else
        //$$ this.button.x = middle - 100 + this.x;
        //$$ this.button.drawButton(Client.getMinecraft(), mouseX, mouseY, partialTicks);
        //#endif

        super.draw(mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int) {
        if (this.hidden) return

        if (this.button.mousePressed(Client.getMinecraft(), mouseX, mouseY)) {
            this.value = (!this.value)
            this.button.playPressSound(Client.getMinecraft().soundHandler)
        }

        if (this.resetButton.mousePressed(Client.getMinecraft(), mouseX, mouseY)) {
            this.value = this.defaultValue
            this.resetButton.playPressSound(Client.getMinecraft().soundHandler)
        }

        this.button.displayString = stringValue
    }
}